### PR TITLE
Add match feature

### LINKED ISSUE #24: Matching feature
As a user, I want to match my Clients (buyer with seller) so that I can spot if there are any Property being sold by a Seller that a Buyer has a preference for.

### PR DESCRIPTION
**Pull request by cindyangXQ**
In this version, 
- each person has at most one preference
- `match` only supports the case where each person has at most 1 property and at most 1 preference
- editing preference is not supported
- tests and documentation not completed

Todo: 
- [X] merge conflict
- [X] `address` should not be required equal in `Person.matches()`
- [ ] update UI of `MatchWindow`

This PR closes #24.

### CODE REVIEW

**Comment by waffledood** [comment]
Hi Cindy, do add the line "closes #id" (where id is the issue number of your feature) in your PR description, so that upon merging your PR, the issue corresponding to your feature gets closed as well

File: src/main/java/seedu/address/commons/core/Pair.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package seedu.address.commons.core;
+
```
Comment (hsiaojietng) [inline src/main/java/seedu/address/commons/core/Pair.java:2]: Would be good to add comments on what this class is for

File: src/main/java/seedu/address/logic/commands/CommandResult.java [hunk 2]
```diff
@@ -42,6 +46,10 @@ public boolean isShowHelp() {
         return showHelp;
     }
 
+    public boolean isShowMatch() {
+        return showMatch;
```
Comment (hsiaojietng) [inline src/main/java/seedu/address/logic/commands/CommandResult.java:50]: Nicely abstracted, good coding!

File: src/main/java/seedu/address/logic/commands/MatchCommand.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package seedu.address.logic.commands;
+
+import static java.util.Objects.requireNonNull;
+
+import seedu.address.model.Model;
+
+public class MatchCommand extends Command {
```
Comment (hsiaojietng) [inline src/main/java/seedu/address/logic/commands/MatchCommand.java:7]: Maybe add a header comment on what this command does

File: src/main/java/seedu/address/logic/parser/CliSyntax.java [hunk 4]
```diff
@@ -12,5 +12,6 @@
     public static final Prefix PREFIX_ADDRESS = new Prefix("a/");
     public static final Prefix PREFIX_TAG = new Prefix("t/");
     public static final Prefix PREFIX_PROPERTY = new Prefix("pr/");
+    public static final Prefix PREFIX_PREFERENCE = new Prefix("pf");
```
Comment (hsiaojietng) [inline src/main/java/seedu/address/logic/parser/CliSyntax.java:15]: Suggestion: Better if this follows the "pf/" format

**Review by hsiaojietng — APPROVED** [review]
Just some nits to check, other than that, LGTM

File: src/main/resources/view/PersonListCard.fxml [hunk 5]
```diff
@@ -9,6 +9,7 @@
 <?import javafx.scene.layout.Region?>
 <?import javafx.scene.layout.VBox?>
 
+<?import com.sun.javafx.charts.Legend?>
```
Comment (waffledood) [inline src/main/resources/view/PersonListCard.fxml:12]: just wondering, is `Legend` being used in `PersonListCard.fxml`? Don't seem to see it here

**Review by waffledood — COMMENTED** [review]
Code looks fine, but take note of merge conflicts you'll need to resolve

Do update & resolve the merge conflicts before proceeding with merging into master

File: src/main/java/seedu/address/logic/parser/CliSyntax.java (hunk 4 above)
Comment (cindyangXQ) [reply src/main/java/seedu/address/logic/parser/CliSyntax.java]: Careless mistake. Resolved!

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java [hunk 6]
```diff
@@ -45,13 +46,17 @@ public AddCommand parse(String args) throws ParseException {
         Phone phone = ParserUtil.parsePhone(argMultimap.getValue(PREFIX_PHONE).get());
         Email email = ParserUtil.parseEmail(argMultimap.getValue(PREFIX_EMAIL).get());
         Address address = ParserUtil.parseAddress(argMultimap.getValue(PREFIX_ADDRESS).get());
-        // property is optional since it should not be present if the person being added is a buyer
+        // property is optional since it should not be present if the person being added is a not a seller.
         Optional<String> propertyArg = argMultimap.getValue(PREFIX_PROPERTY);
         Optional<Property> property =
                 propertyArg.isPresent() ? Optional.of(ParserUtil.parseProperty(propertyArg.get())) : Optional.empty();
+        // property is optional since it should not be present if the person being added is a not a buyer.
```
[the remark this answers is not in the document]
Comment (cindyangXQ) [reply src/main/java/seedu/address/logic/parser/AddCommandParser.java]: Resolved!

File: src/main/java/seedu/address/model/person/Person.java [hunk 7]
```diff
@@ -81,6 +88,19 @@ public boolean isSamePerson(Person otherPerson) {
...
+     */
+    public boolean matches(Person preference) {
+        if (property.isEmpty()) {
+            return false;
+        }
+        if (preference.preference.isEmpty()) {
+            return false;
+        }
+        return property.get().equals(preference.preference.get());
+    }
```
[the remark this answers is not in the document]
Comment (cindyangXQ) [reply src/main/java/seedu/address/model/person/Person.java]: Good suggestion. Changed to `buyer`!

File: src/main/resources/view/PersonListCard.fxml (hunk 5 above)
Comment (cindyangXQ) [reply src/main/resources/view/PersonListCard.fxml]: Careless mistake. Resolved!

File: src/main/java/seedu/address/model/person/Person.java [hunk 8]
```diff
@@ -25,18 +25,21 @@
     // Data fields
     private final Address address;
     private final Optional<Property> property;
+    private final Optional<Property> preference;
```
Comment (Ashuh) [inline src/main/java/seedu/address/model/person/Person.java:28]: Is preference supposed to represent a specific `Property` that the client wants to buy or the general criteria of the `Property` that needs to be satisfied?

**Review by waffledood — APPROVED** [review]
Updated changes look good. LGTM!

**Review by howardwhw2 — APPROVED** [review]
LGTM. No issues found.

**Pull request merged**